Inserter: Resize canvas to fit when pattern category selected

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
One of the things we have learnt from zoom out mode is that its very annoying when the pattern category previews overlap the canvas. This updates the canvas to resize based on whether the pattern category previews panel is open.

## Why?
It's annoying to have this overlapping the canvas.

## How?
Create a new private value in the store to save whether or not the sidebar is wide. Ideally the sidebar should be able to be any size but that's out of the scope of this PR.

## Testing Instructions
1. Open the Site Editor
2. Open the inserter
3. Select the patterns tab
4. Select a pattern category
5. Notice that the canvas resizes when the category previews are open, so that they don't overlap the content.

## Screenshots or screencast <!-- if applicable -->
<img width="1715" alt="Screenshot 2024-07-26 at 12 54 25" src="https://github.com/user-attachments/assets/e3c0e3c5-94ef-48dd-abc2-9cc077690d36">

scruffian committed:
Inserter: Make the inserter wider when the pattern category is selected

scruffian commented on packages/interface/src/components/interface-skeleton/index.js:
```diff
@@ -19,6 +19,8 @@ import {
 	useResizeObserver,
 } from '@wordpress/compose';
+import { useSelect } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
```
I don't think we're allowed to use the block editor package inside interface 🤔

jeryj replied:
Probably not -- maybe we shouldn't be storing the pattern category. As in, we don't really care that we have a pattern category selected, we want to tell the interface how to be laid out. So, something like 'wide-sidebar' or 'zoom-out-wide'?

scruffian replied:
good idea!

scruffian commented on packages/interface/src/components/interface-skeleton/index.js:
```diff
@@ -182,14 +197,15 @@ function InterfaceSkeleton(
 									closed: { width: 0 },
 									mobileOpen: { width: '100vw' },
+									patternsOpen: {
+										width: '650px',
```
300 + 350 - should be a variable

scruffian commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -87,6 +87,10 @@ $block-inserter-tabs-height: 44px;
 }
 
+.block-editor-inserter__menu.show-panel {
+	width: 650px;
```
Should be a variable

scruffian commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -61,9 +61,19 @@ function InserterMenu(
+				select( blockEditorStore ).selectedPatternCategory();
+			return _selectedPatternCategory
+				? _selectedPatternCategory
+				: __experimentalInitialCategory;
```
Does this work as a fallback?

scruffian commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -299,7 +303,7 @@ $block-inserter-tabs-height: 44px;
 		border-left: $border-width solid $gray-200;
 		padding: 0;
-		left: 100%;
+		right: 0;
```
What impact does this have on the mobile view?

jeryj commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -61,9 +61,19 @@ function InserterMenu(
+		},
+		[ __experimentalInitialCategory ]
 	);
+	const { setSelectedPatternCategory } = useDispatch( blockEditorStore );
```
To explain my other comment more: instead of dispatching `setSelectedPatternCategory` to the block store, maybe dispatch something to the interface package about the layout we want. In the interface package we don't care what category is selected. We care about the layout we want when we have a category selected.

scruffian committed:
move to interface store

scruffian committed:
Use variables

scruffian committed:
use variable

scruffian committed:
revert unused changes

scruffian committed:
remove unconnected changes

scruffian committed:
move display flex for all inserter menu instances

scruffian committed:
import internally

scruffian committed:
lock the new selector and action

scruffian commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -82,11 +82,16 @@ $block-inserter-tabs-height: 44px;
 }
 
 .block-editor-inserter__menu {
+	display: flex;
```
There is a risk that this breaks something

scruffian commented:
The one problem I have with this PR is this error:
`Store "core/interface" is already registered.`

richtabor commented:
Just confirming, but this is assuming the zoom out with patterns experiment is disabled, correct?

richtabor commented:
It's certainly better than cover up the content. Nice.

scruffian commented:
> Just confirming, but this is assuming the zoom out with patterns experiment is disabled, correct?

Yes

Commit:
remove shadow from category panel

no longer needed with this layout adjustment, as the .interface-interface-skeleton__secondary-sidebar maintains the shadow.

Commit:
resolve inserter on tablet viewport

Commit:
only apply width on < medium

Commit:
Use $sidebar-width instead of $category-panel-width

richtabor commented:
Pushed a few small design tweaks, removing the duplicate shadow (now that its not needed for this layout), resolving the flex width on mobile, and using $sidebar-width instead of a new variable (buys a little space with a -20px difference).

richtabor commented:
Curious, does this make https://github.com/WordPress/gutenberg/issues/63510 easier to resolve perhaps?

richtabor reviewed (commented):
Design wise, this is good to go. 


### Trunk

https://github.com/user-attachments/assets/861ec268-95a0-4dde-ab1c-238026d1c5b0


### This PR

https://github.com/user-attachments/assets/42a1f60f-02ec-4554-b999-bb39cc4d8130

jeryj commented on packages/block-editor/src/components/tabbed-sidebar/style.scss:
```diff
@@ -6,7 +6,7 @@
 	overflow: hidden;
 
 	@include break-medium() {
-		width: 350px;
```
I think a simpler plan will be to:
- When pattern categories are open
- Update this width to be the desired total width
- Remove all the absolute hacks for the pattern categories

scruffian commented:
This was resolved by https://github.com/WordPress/gutenberg/pull/64110
